agent/acs/handler/payload_responder.go

```diff
@@ -0,0 +1,289 @@
+				credentials.ApplicationRoleType)
+			err = tmHandler.credentialsManager.SetTaskCredentials(
+				&(credentials.TaskIAMRoleCredentials{
+					ARN:                aws.StringValue(task.Arn),
+					IAMRoleCredentials: taskIAMRoleCredentials,
+				}))
+			if err != nil {
+				tmHandler.handleUnrecognizedTask(task, err, payload)
+				allTasksOK = false
+				continue
+			}
+			apiTask.SetCredentialsID(taskIAMRoleCredentials.CredentialsID)
+		}
+
+		// Add ENI information to the task struct.
+		for _, acsENI := range task.ElasticNetworkInterfaces {
+			eni, err := apieni.ENIFromACS(acsENI)
+			if err != nil {
+				tmHandler.handleUnrecognizedTask(task, err, payload)
+				allTasksOK = false
+				continue
+			}
+			apiTask.AddTaskENI(eni)
+		}
+
+		// Add the app mesh information to task struct.
+		if task.ProxyConfiguration != nil {
+			appmesh, err := apiappmesh.AppMeshFromACS(task.ProxyConfiguration)
+			if err != nil {
+				tmHandler.handleUnrecognizedTask(task, err, payload)
+				allTasksOK = false
+				continue
+			}
+			apiTask.SetAppMesh(appmesh)
+		}
+
+		if task.ExecutionRoleCredentials != nil {
+			// The payload message contains execution credentials for the task.
+			// Add the credentials to the credentials manager and set the
+			// task executionCredentials id.
+			taskExecutionIAMRoleCredentials := credentials.IAMRoleCredentialsFromACS(task.ExecutionRoleCredentials,
+				credentials.ExecutionRoleType)
+			err = tmHandler.credentialsManager.SetTaskCredentials(
+				&(credentials.TaskIAMRoleCredentials{
+					ARN:                aws.StringValue(task.Arn),
+					IAMRoleCredentials: taskExecutionIAMRoleCredentials,
+				}))
+			if err != nil {
+				tmHandler.handleUnrecognizedTask(task, err, payload)
+				allTasksOK = false
+				continue
+			}
+			apiTask.SetExecutionRoleCredentialsID(taskExecutionIAMRoleCredentials.CredentialsID)
+		}
+
+		validTasks = append(validTasks, apiTask)
+	}
+
+	// Add 'stop' transitions first to allow seqnum ordering to work out
+	// Because a 'start' sequence number should only be proceeded if all 'stop's
+	// of the same sequence number have completed, the 'start' events need to be
+	// added after the 'stop' events are there to block them.
+	stoppedTasksCredentialsAcks, stoppedTasksAddedOK := tmHandler.addTasks(payload, validTasks, isTaskStatusNotStopped)
+	newTasksCredentialsAcks, newTasksAddedOK := tmHandler.addTasks(payload, validTasks, isTaskStatusStopped)
+	if !stoppedTasksAddedOK || !newTasksAddedOK {
+		allTasksOK = false
+	}
+
+	// Construct a slice with credentials acks from all tasks.
+	credentialsAcks := append(stoppedTasksCredentialsAcks, newTasksCredentialsAcks...)
+	return credentialsAcks, allTasksOK
+}
+
+// handleUnrecognizedTask handles unrecognized tasks by sending 'stopped' with
+// a suitable reason to the backend.
+func (tmHandler *taskMessageHandler) handleUnrecognizedTask(task *ecsacs.Task, err error,
```

Comment:
The function name handleUnrecognizedTask seems a bit of a misnomer. As I understand from the function above, this is being called in a lot more cases than just unrecognized task payload. Maybe a better name would be "handleInvalidTaskPayload" or "handleErrorInTaskPayload". Just a suggestion. :)

Reply:
I believe the function is called whenever performing a specific action on a task associated with an ACS payload message errors out. So I think the function name assumes such tasks are considered "unrecognized". I agree that "invalid" is probably a better word choice here. If it works with you, I can update the function name accordingly in the next revision if any blocking changes need to be made. 👍

Reply:
Sure. Sounds good. It might be good to track as part of TODOs in the PR for future reference.